Multiple, simultaneous SPI handling and Smart SD mounting - test

LinoBarreca opened the pull request:
We will use this place to have a feedback from testers.
**_SKR-PRO ONLY_**
(other boards are broken at the moment, don't waste your time trying to build it, also, I temporarily disabled them and deleted config files from this branch since they slow down the IDE)

We need to test the following situations:
Printer without any sd card inserted.
Screen with sd-slot and card **in screen** sd-slot.
Screen with sd-slot and card on board.
Screen without sd-slot and card on board.

at the moment `LULZBOT_TOUCH_UI` and `MAX6675_SEPARATE_SPI` are broken.
if you have those DON'T use this version.

How to test:
Pull the sources.
Add your config files.
Compile.
Put the firmware in the onboard sd.
Put a .gcode in the onboard sd.
Turn on the printer.
Check that you can see the gcode.

LinoBarreca committed:
Handle multiple SPI busses
..and use first sd-card found at boot.

LinoBarreca committed:
Add serial output for debug

LinoBarreca committed:
HAL_init sets SS lines high on every bus

LinoBarreca committed:
Test only SKR-PRO for now (temporary)

LinoBarreca committed:
Fix build when SD_CHECK_AND_RETRY is active

LinoBarreca committed:
Fix build when SD_CHECK_AND_RETRY is not active

LinoBarreca committed:
Merge branch 'OnBoardSD' of https://github.com/LinoBarreca/Marlin into OnBoardSD

LinoBarreca committed:
Added better comments

LinoBarreca committed:
Change SD_DETECT pin according to SPI BUS used

LinoBarreca committed:
Check correct pin to detect card

LinoBarreca committed:
Use same bus for all functions during init

LinoBarreca committed:
Fix for rushed commit while drunk. Sorry.

LinoBarreca committed:
Temporary fix for blank screen

LinoBarreca committed:
Proper pins in variant

LinoBarreca committed:
I might be wrong...but...

LinoBarreca committed:
Rollback last commit (test failed) and set NOPULL
(since it's done in Hardware through R1,2,5,10 and 12)

LinoBarreca committed:
Explicitly disable unused HAL drivers
(to avoid conflicts on pin usage)

LinoBarreca committed:
Some more safe disables.

LinoBarreca committed:
Fix for serial ports(we have one more now) and USB

LinoBarreca committed:
More PIN cleaning and clarification

LinoBarreca committed:
Comments on pins...Still have to check USB HOST

LinoBarreca committed:
Revert "Merge from LinoBarreca/NewTimer"

This reverts commit cca921198df71e096605e018202b0cf5cb806565, reversing
changes made to 59b5d6a8181bd4282879f04b9a94afb8e08c250b.

thinkyhead committed:
"nozzle_to_probe_offset" is obsolete

thinkyhead committed:
Tweak platformio.ini alignment

Evg33 commented:
7a28f4ea09871e9b852aebe1174d6f5e2eaf300f

```
echo:Searching for SD card on SPI bus
0
echo:Searching for SD card on SPI bus
1
echo:Searching for SD card on SPI bus
2
echo:SD init fail

```

LinoBarreca commented:
I'm not working on this now. A guy told me he had problems with fans. I am fixing that.

LinoBarreca committed:
Enable clock on the HAL

LinoBarreca commented:
try now, please ;)

akselii commented:
Boots fine, but

echo:Searching for SD card on SPI bus
0
echo:Searching for SD card on SPI bus
1
echo:Searching for SD card on SPI bus
2
echo:SD init fail

LinoBarreca commented:
@akselii Okay. This makes me think the problem is in Marlin core.
Let me explain the thinking behind:

We know that the card is working (not a hardware problem) because the bootloader from BTT is working: it initializes the hardware (not just the SPI on TF, also clocks and more stuff), flashes the code correctly, renames the file and then boots...so the hardware is working.

What then could possibly prevent marlin to read the card?
1) marlin has the wrong pins so that it's trying to talk on wrong ports (and we saw yesterday this is not true with the continuity tests) so marlin has the correct pins.
marlin says to STM SPI library to init the bus on thos pins. The SPI library doesn't know how to initialize them so it opens the PinMap_SPI_* defined in buildroot\share\PlatformIO\variants\BIGTREE_SKR_PRO_1v1\PeripheralPins.c gets the first occurrence of the pin, and sets the hardware according to what's written in there.
2) since the STM SPI library configures the pins as "first found" and we had (wrong) multiple definitions for the same pin maybe the hal initialized the hardware on the wrong pins. I fixed every pin conflict and now they are unique. so that can't be the reason. the STM SPI library configures the hardware on the correct pins.
3) maybe it configures them wrong assuming they are active because the bootloader, kinda logically, after reading the card deactivates the clock when it has finished using it. but this isn't possibile since, with the last commit I init the hardware devices from zero. so the SPI library actually configures the pins which have the underlying hardware correctly activated. so the hardware should be activated correctly. the hardware should work properly and it should have been properly initialized and configured....then marlin tries to talk with the card.....and shit happens....
I wanna see if I can slow down, in marlin, that talk....maybe it's too fast for the cards on that (very fast) cortex. I'm pushing another commit. wait

LinoBarreca committed:
Slow down SPI speed (test)

LinoBarreca commented:
try again please :)

akselii commented:
Same as before, 0, 1, 2 init fail. If i insert a card to the screen it finds it fine.

LinoBarreca commented:
tomorrow we will try something completely different. good night for now.

thinkyhead committed:
[cron] Bump distribution date

LinoBarreca commented:
Guys. I have big news on this.
I passed all the night awake studying, on paper, on the circuit schematics what was wrong..and I'm pretty sure I found where the problem is.
I need @sjasonsmith (the only one which has the oscilloscope here) to make a test for me:
1) Take the board only. no need to be connected to anything.
2) build the version that I will push after this comment . It's an empty (= doing nothing) sketch but big in size. "big" because we need the bootloader to read and flash it (if it's too fast because the sketch is too small you won't be able to see what I suppose it's happening)
3) Put it in the tf card.
4) Attach an oscilloscope to pin 5 on the TF
![image](https://user-images.githubusercontent.com/897723/68676690-cfd10900-055a-11ea-900c-a4cc064552ed.png)
5) Power the board. If I am right you will **NOT** see a clock on the pin 5 (TF_CLK) but rather a floating, not LOW, signal. After the bootloader is gone that pin 5 should go HIGH (because pulled high by resistors)
6) leaving everything as is now probe TF pin 2 (TF_SS). It should be high.
7) keeping the probe on PIN 2 push and release reset button. the bootloader should start and you should see the clock on PIN 2 lasting for some seconds, while the bootloader reads the card.
After the bootloader is gone that should go HIGH (because pulled high by resistors)

If I was right I have the solution and the explanation.

LinoBarreca committed:
Do-nothing Marlin

LinoBarreca committed:
No setup either.

sjasonsmith commented:
Here is both lines while loading the firmware from the card.
![Pin2_and_5](https://user-images.githubusercontent.com/20053467/68847242-318c9280-0683-11ea-8de1-a90fb440c72c.png)

LinoBarreca commented:
@sjasonsmith thank you. can you zoom in, please?
here.
![image](https://user-images.githubusercontent.com/897723/68848282-5529fa00-06d0-11ea-870b-a2a5c256910b.png)

BTW: 7 seconds of boot is A LOT

tpruvot committed:
STM32F1: Remove extra py imports (#15898)

LinoBarreca commented:
(That's half proof) I want to check if the yellow is a clock or floating.

thinkyhead committed:
Fix timers.h include

LinoBarreca committed:
Don't compile STM32 SoftwareSerial unless STM32 (#15894)

LinoBarreca committed:
SKR-PRO: Fix FAN0 PWM speed with Heater2 (#15889)

KiteLab commented on Marlin/src/pins/stm32/pins_BIGTREE_SKR_PRO_V1.1.h:
```diff
@@ -186,7 +186,7 @@
 // MOSI, MISO, SCK , SS  , SW
   {PB15, PB14, PB13, PB12, PF12}, //EXT2 (Probed first. The card in the display it's probably easier to reach.)
-  {PB5 , PA6 , PA5 , PA4 , PB11}, //TF
+  {PA7 , PA6 , PA5 , PA4 , PB11}, //TF
```
Seems to be plausible. D13 at PA7 is flashing during firmware updates.
(If we can still have any trust in the SCH)

LinoBarreca replied:
No. @akselii did with me some continuity tests. It's definitely pb12.
Probably pa7 flashes because they instructed the firmware to.
When I will have the board here I will do all the measurements myself.

thinkyhead committed:
Reduced string storage using tokens (#15593)

akselii commented:
Oh yeah i think the waveform sjasonsmith provided is when the bootloader is flashing a firmware to the chip which is why its taking 7 secs. The normal boot time for the board is fast.

sjasonsmith commented:
> Oh yeah i think the waveform sjasonsmith provided is when the bootloader is flashing a firmware to the chip which is why its taking 7 secs. The normal boot time for the board is fast.

Yes, this is a full firmware load from the flash, using the commits @LinoBarreca pushed.
I'll check again in a while to see if I can confirm whether those signals are driven or if we are just seeing crosstalk.

LinoBarreca commented:
Only if you want @sjasonsmith. Another meaningful comparison would be polling the same lines while trying to access the other spi devices. The way the card is used (red line) makes me think they aren't using the spi but rather a bit banging.
(Which makes sense since the bootloader is written in native stm32 code)
I guess they didn't have the help of the available libraries and they just bit bang the lines. If that's what happens (what I think) the jellow line (the clock) should not be "perfect" (a hardware spi takes the clock from the internal circuitry) a "software spi" is subject to normal software delays according to the instructions you di between high and low.
Studying the papers I found that there's a problem in how they choose the lines.
If you setup your "devices", inside the chip, you can't connect them all. Some go connected in a way, other in another way. If you use the clock source for, let's say a timer, and you need another, different frequency, on another device (spi) it might be possible or not according on which function you assign to which port.
So, yesterday night I found that error.
I woke up, asked you to do that measurement and, waiting for the zoom, I opened the chip designer and did the same configuration that they did on the pinout.
Around 17 I had "configured" the "new chip" with the ports that btt choose...and guess what? There's a warning about the clock used in a wrong way. The designer offers to "fix" ports for you. But only if you let him choose which port to use. So it can correctly calculate clocks and dividers to each port.

Since the hardware clock can't work with current pinout I suspected it was floating (= undefined) but they do access the card.
How? They only have to read a small firmware file 1mb max. They most likely bitbang it.
Which is easy for reading a file, but killing performances for reading a gcode at printing speeds.
I have a couple of solutions but I first need to see if that's a hardware clock or a software clock.
They had two possibilities:
1) If it's a hardware clock (= perfect wave at a fixed frequency) it means that they are using hardware spi and, in the bootloader they only initialize the circuitry making the stm work as it only had one spi port and nothing else. (Enough to read card and flash it). When they are done with it they deinitialize the spi and initialize all the other circuitry.
2) If that's a software clock (= floating) they had the same warning in the chip designer and they thought "okay, let's initialize the test of the hardware correctly and bitbang the tf"

sjasonsmith commented:
Here are some new pictures for you. I biased both lines to ground with a 4.7k resistor, when they are floating you will see them hover at ~1.25V. It is clear that both are driven when signals are seen, and both are floating at the end of the full firmware load.

**Reset, no card present**
![Reset_Biased_Low](https://user-images.githubusercontent.com/20053467/68923581-80920080-0733-11ea-951e-9e53682f369d.png)

**Reset, no card present, zoomed on one of the narrow pulses**
![Reset_Biased_Low_ZoomedOnPulse](https://user-images.githubusercontent.com/20053467/68923621-94d5fd80-0733-11ea-9232-506fdd3ccd2d.png)

**Firmware load, full image (some gaps are due to aliasing)**
![FirmwareLoad_Biased_Low](https://user-images.githubusercontent.com/20053467/68923652-a5867380-0733-11ea-8d5b-50c54f4b8135.png)

**Firmware load, zoomed to the center section you requested (Look at top bar to see placement in waveform)**
![FirmwareLoad_Zoomed](https://user-images.githubusercontent.com/20053467/68923737-db2b5c80-0733-11ea-9967-d397f23b627d.png)

**Same image as above, zoomed further to show one of the falling edges**
![FirmwareLoad_Zoomed2](https://user-images.githubusercontent.com/20053467/68923776-f5653a80-0733-11ea-95be-45311ed022f3.png)

LinoBarreca commented:
Sorry. With floating I didn't mean floating in levels but rather floating in frequency.
![20191115_081439](https://user-images.githubusercontent.com/897723/68924186-7a764100-0780-11ea-8be3-d595ad2008c2.jpg)
That's what I was searching for.

LinoBarreca commented:
If you have an analyzer on your scope and you can check the frequency from point A to point B, you will see that the frequency of the yellow line varies (even if considered on the same "area" lenght, like 10ms) that should mean it doesn't come from an hardware clock

LinoBarreca commented:
Also...can you zoom more (horizontally) in the same place I underlined? That doesn't even look like a square wave.

LinoBarreca commented:
This doesn't even look like a square wave to me
![Screenshot_20191115-083240_Firefox](https://user-images.githubusercontent.com/897723/68925080-895df300-0782-11ea-855c-6732c0f003ff.jpg)

sjasonsmith commented:
The quality of my analog capture is not very good. I don't do this often, so I don't know how to get very reliable measurements. People at work are always talking about how important expensive probes are which I certainly do _not_ have. Of course they are probably probing PCIe or DDR4 signals.  :smile:


At the highest zoom level the yellow signal is very noisy. The transitions seem to bounce a lot. I don't know if that is real or bad probing. At anything less than the highest zoom there is lots of aliasing in the image, and you can't really trust the shapes you see.

LinoBarreca commented:
I think it's real.and that's why they use the SS.
BASICALLY that's how it should be
From the master (board)
Constant clock on 5 from init to end.
Constant low on 2 to tell the TF "from now on I'm talking with you"
Bit sent on mosi when clock rises or falls (according to protocol)
On the slave (card)
If 5 and 2 store bit from miso.

What is happening now:
5 isn't good. They just transmit  USING THE SS as clock on a 5 which isn't good (probably the error I got from the chip designer) so they keep the
"I'm talking with you" (pin 2)and the mosi  it at the desired value for enough time for the card to catch a "random" high. So storing the bit.
Then they clear ss, like the master is talking to some other chip.
Next bit:
Set mosi at desired level, set ss, wait, unset ss....and so on.

thinkyhead committed:
Update TMC_ADV examples

thisiskeithb committed:
Support 512K "RCT6" in BTT SKR Mini (#15890)

Roxy-3D committed:
Eliminate multiple settings to extruder_duplication_enabled in G28

If the IDEX mode was previously set (via GCode or the LCD Panel) to a duplication mode...   It was lost after a G28 due to this redundant setting of 'extruder_duplication_enabled'. 

It is very important to the various use cases of IDEX machines that extruder_duplication_enabled is preserved across a G28.   The reason is this allows a person to use the LCD Panel to put the machine into a duplication mode, and then send a 'normal' part file to the printer and get two parts out of the process.

thinkyhead committed:
Only TMCStepper 0.5.x is certified

robbycandra committed:
Use lcd_uint_t in lcd_put_u8str_ind_P (#15909)

thinkyhead committed:
[cron] Bump distribution date

thinkyhead committed:
[cron] Bump distribution date

LinoBarreca committed:
Handle multiple SPI devices simultaneously
and allow to read and write to each of them targeting the desired one.

Re-enable code disabled for test
Introduce the possibility to do SoftwareSPI (disabled for now,needs lib)
Upgrade TMCStepper

LinoBarreca commented:
Do not build next commit, please. It isn't meant to be tested yet.

LinoBarreca committed:
Cleaning of definitions and macros

Added SoftSPI (still disabled for now, gotta check it yet)

swilkens committed:
TMCStepper 0.6.x is now Marlin-compatible (#15950)

shitcreek committed:
Update Vietnamese language (#15940)

InsanityAutomation committed:
Match onMeshUpdate() index sign (#15920)

TheAssassin committed:
Ender-2 LCD on BTT SKR E3 (#15924)

GMagician committed:
Fix MMU2 compile error (#15923)

sjasonsmith committed:
Update Anycubic Kossel config (#15935)

Evgeny-SPB committed:
G34 logic / typo fix (#15938)

sjasonsmith committed:
Allow UBL G29 J1 with PAUSE_BEFORE_DEPLOY_STOW (#15934)

GMagician committed:
Update Italian language (#15925)

InsanityAutomation committed:
Float XY probe offset followup (probe edge) (#15928)

danandrei96 committed:
Add Zonestar/Anet LCD for SKR mini E3 (#15931)

thisiskeithb committed:
SKR Mini E3 1.0/1.2 configs (#15941)

InsanityAutomation committed:
Better singlenozzle stored fan speed menus (#15921)

thinkyhead committed:
Followup to meshCount patch

bigtreetech committed:
Power Loss Recovery with a UPS (#15943)

rmoravcik committed:
Update Slovak language (#15953)

thinkyhead committed:
Update thermocouple comments

thinkyhead committed:
Drop obsolete setting

bob808 committed:
MAX31865 temperature sensor (#15930)

sjasonsmith committed:
Fix G29 for updated meshCount type (#15960)

sjasonsmith committed:
Improve Delta probing / calibration (#15887)

thinkyhead committed:
Bring configs up to date

thinkyhead committed:
Update sensor -5 label

sl1pkn07 committed:
Support RET6 in BTT SKR Mini (#15957)

LinoBarreca committed:
New detection logic for SD cards

LinoBarreca committed:
Merge 'upstream/bugfix-2.0.x' into MultiSPI

sjasonsmith commented:
@LinoBarreca did you ever get your board? I see you’ve started making edits again.

LinoBarreca commented:
> @LinoBarreca did you ever get your board? I see you’ve started making edits again.

not yet. but I'm ironing anything that could cause problems.
Added some debug output and asked on fb if someone can give me the serial output of current code (to see what happens while I wait for the board..so I try to track the problem by looking at the debug output messages)

LinoBarreca commented:
Opened another pull request (to myself) since github closed this when I force-pushed the branch after a rebase

https://github.com/LinoBarreca/Marlin/pull/7
